### PR TITLE
Add communication between client and server, send packet from client to server and wait for return the same package

### PR DESCRIPTION
**Pull request by ghost**
make public NetworkReader (class for read from stream), NetworkWriter(class for write to stream), NetworkStreams(class for get Inputstream/Outputstream from socket)
add AppClientCommunicationHandler - method 'connect' with server and set object for communication (NetworkReader, NetworkWriter), method 'write' write Packet to server and method 'read' read Packet from server

### CODE REVIEW

File: client/src/main/java/com/blackship/battlesheep/communication/network/AppClientCommunicationHandler.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+    void connect() throws IOException {
+        client.connect();
+
+        Streams clientStreams = new NetworkStreams(client.getSocket());
+
+        InputStream clientInputStream = clientStreams.getInput()
+                .orElseThrow(() -> new IOException("Null inputstream !"));
+        OutputStream clientOutputStream = clientStreams.getOutput()
+                .orElseThrow(() -> new IOException("Null outpustream !"));
```
Comment (bartwoj) [inline client/src/main/java/com/blackship/battlesheep/communication/network/AppClientCommunicationHandler.java:38]: lambdas 👍

File: client/src/main/java/com/blackship/battlesheep/communication/network/AppClientCommunicationHandler.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    void write(Packet packet) throws IOException {
+        clientWriter.write(new NetworkPacketConverter().toByte(packet));
+    }
+
+    Packet read() throws IOException, ClassNotFoundException {
+        return new NetworkPacketConverter().toPacket(clientReader.read());
+    }
+
+}
```
Comment (bartwoj) [inline client/src/main/java/com/blackship/battlesheep/communication/network/AppClientCommunicationHandler.java:52]: Separate class for managing client connection 👍

File: client/src/test/java/com/blackship/battlesheep/communication/network/AppClientSocketIT.java [hunk 3]
```diff
@@ -1,36 +1,56 @@
...
+                receivedPacketIsTheSameAsSent = EqualsBuilder.reflectionEquals(givenPacketToSend, receivedPacket);
+            } catch (IOException | ClassNotFoundException | InterruptedException e) {
+                e.printStackTrace();
+                receivedPacketIsTheSameAsSent = false;
             }
-        };
-        new Timer().schedule(connectTask, 500);
-        givenServerSocket.accept();
+        });
+        connectAndHandle.start();
```
Comment (bartwoj) [inline client/src/test/java/com/blackship/battlesheep/communication/network/AppClientSocketIT.java:43]: Optional: you can start() thread on the same line as you end its declaration " ...}).start(); ", but overall 👍

File: client/src/test/java/com/blackship/battlesheep/communication/network/AppClientSocketIT.java [hunk 4]
```diff
@@ -1,36 +1,56 @@
...
+
+        Socket givenClientSocket = givenServerSocket.accept();
+
+        byte[] buffer = new byte[2048];
```
Comment (bartwoj) [inline client/src/test/java/com/blackship/battlesheep/communication/network/AppClientSocketIT.java:47]: "magic number" - extract it to some MAX_BUFFER_CAPACITY or something.

File: pom.xml [hunk 5]
```diff
@@ -70,6 +71,11 @@
             <artifactId>slf4j-log4j12</artifactId>
             <version>${slf4jVersion}</version>
         </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>${commonLang3Version}</version>
+        </dependency>
```
Comment (bartwoj) [inline pom.xml:78]: 👍

File: common/src/main/java/com/blackship/battlesheep/common/communication/network/NetworkReader.java [hunk 6]
```diff
@@ -19,8 +19,9 @@
 
         ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
 
-        for(int inputValue = inputStream.read(); inputValue != -1 || inputStream.available() < 0 ; inputValue = inputStream.read()) {
+        for(int inputValue = inputStream.read(); inputValue != -1; inputValue = inputStream.read()) {
             byteArrayOutputStream.write(inputValue);
+            if (inputStream.available() <= 0) break;
```
Comment (bartwoj) [inline common/src/main/java/com/blackship/battlesheep/common/communication/network/NetworkReader.java:24]: Optional: better is strong inequality like "< 1"

**Review by bartwoj — APPROVED** [review]
Consider this minor changes further in, but overall it is approved.

**Review by arturAR — APPROVED** [review]

**Pull request merged**